rfcs/0019-storing-client-data.md

```diff
@@ -0,0 +1,490 @@
+
+## SQL Database
+
+This architecture supports many SQL databases and it can even exclude a database entirely to make the architecture modular. When it is included, there are different use cases for using a light database, such as SQLite, and heavier database, such as Postgres. Postgres or even a distributed cloud database, such as AppSync, should be used when we have an archive node or a block explorer trying to make efficient writes to store everything that it hears from the network. These databases also support built-in subscriptions that would be convinient for clients to listen to.
+
+For the objects discussed in the previous section, we can embed them as tables in one global SQL database. Below are the schemas of the tables:
+
+```
+Table block {
+  state_hash string [pk]
+  creator string [not null]
+  staged_ledger_hash string [not null]
+  ledger_hash string [not null]
+  epoch int [not null]
+  slot int [not null]
+  ledger_proof_nonce int [not null]
+  status int [not null]
+  block_length int [not null]
+  block_time date [not null]
+  Indexes {
+    (block_length, epoch, slot) [name: "block_compare"]
+    block_time [name: "block_compare"]
+  }
+}
+
+  // Consensus_status is an ADT, but with its small state spaces, it can be coerced into an integer to make it easy to store it in a database. Here are the following states:
+  // - pending : The block is in the transition frontier and its value is its block confirmation number. Value[0, (k-1)]
+  // - confirmed : The block has reached finality by passing the root. Value=-1
+  // - failure: The block got removed from the transition frontier. Value=-2
+  // - unknown: After becoming online and bootstrapping, we may not know what will happen to the block. Value=-3
+
+Table user_commands {
+  id string [pk]
+  is_delegation bool [not null]
+  nonce int64 [not null]
+  sender string [not null]
+  receiver string [not null]
+  amount int64 [not null]
+  fee int64 [not null]
+  memo string [not null]
+  first_seen date
+  transaction_pool_membership bool [not null]
+  is_added_manually bool [not null]
+  Indexes {
+    (status, sender) [name: "fast_pooled_user_command_sender"]
+    (status, receiver) [name: "fast_pooled_user_command_receiver"]
+    (sender, first_seen) [name: "fast_sender_pagination"]
+    (receiver, first_seen) [name: "fast_receiver_pagination"]
+  }
+}
+
+Table fee_transfers {
+  id string [pk]
+  fee int64 [not null]
+  receiver string [not null]
+  first_seen date
+}
+
+Table block_to_transactions {
+  state_hash string [not null, ref: > block.state_hash]
+  transaction_id int [not null, ref: > user_commands.id, ref: > fee_transfers.id]
+  is_user_command bool
+  sender string
+  receiver string [not null]
```

Comment:
I was thinking about this and I think it would be good to have `sender` and `receiver` fields in this join table because many block related queries require you filter based on some predicate of a sender and receiver of the transactions in the block. For example, I may want to see what are the state hashes of the blocks that involves a certain sender S. This type of query would reduce the need to make a lookup join to the transactions table.